File: src/main/java/info/skyblond/archivedag/service/impl/RedisLockService.java
package info.skyblond.archivedag.service.impl;

import lombok.extern.slf4j.Slf4j;
import org.springframework.integration.redis.util.RedisLockRegistry;
import org.springframework.stereotype.Service;

import java.util.concurrent.locks.Lock;

@Slf4j
@Service
public class RedisLockService {
    private static final long DEFAULT_EXPIRE_UNUSED = 60000L;

    private final RedisLockRegistry redisLockRegistry;

    public RedisLockService(RedisLockRegistry redisLockRegistry) {
        this.redisLockRegistry = redisLockRegistry;
    }

    /**
     * Lock the key. Await if the key not available.
     * Return after get the key.
     */
    public void lock(String lockKey) {
        Lock lock = this.obtainLock(lockKey);
        lock.lock();
    }

    /**
     * Unlock the key.
     */
    public void unlock(String lockKey) {
        try {
            Lock lock = this.obtainLock(lockKey);
            lock.unlock();
            this.redisLockRegistry.expireUnusedOlderThan(RedisLockService.DEFAULT_EXPIRE_UNUSED);
        } catch (Exception e) {
            log.error("Cannot unlock {}", lockKey, e);
        }
    }

    private Lock obtainLock(String lockKey) {
        return this.redisLockRegistry.obtain(lockKey);
    }
}

File: backend/src/main/kotlin/info/skyblond/archivedag/ariteg/service/RedisLockService.kt
package info.skyblond.archivedag.ariteg.service

import info.skyblond.archivedag.ariteg.config.AritegProperties
import org.slf4j.LoggerFactory
import org.springframework.integration.redis.util.RedisLockRegistry
import org.springframework.stereotype.Service
import java.util.concurrent.locks.Lock

@Service
class RedisLockService(
    private val aritegProperties: AritegProperties,
    private val redisLockRegistry: RedisLockRegistry
) {
    private val logger = LoggerFactory.getLogger(RedisLockService::class.java)

    /**
     * Lock the key. Await if the key not available.
     * Return after get the key.
     */
    fun lock(lockKey: String) {
        val lock = obtainLock(lockKey)
        lock.lock()
    }

    /**
     * Unlock the key.
     */
    fun unlock(lockKey: String) {
        try {
            val lock = obtainLock(lockKey)
            lock.unlock()
            redisLockRegistry.expireUnusedOlderThan(
                aritegProperties.meta.lockExpireTimeUnit.toMillis(
                    aritegProperties.meta.lockExpireDuration
                )
            )
        } catch (e: Exception) {
            logger.error("Cannot unlock {}", lockKey, e)
        }
    }

    private fun obtainLock(lockKey: String): Lock {
        return redisLockRegistry.obtain(lockKey)
    }
}
